src/modules/vtol_att_control/pusher.cpp

```diff
@@ -0,0 +1,319 @@
+			// in mc mode
+			_vtol_schedule.flight_mode = MC_MODE;
+			_mc_att_ctl_weight = 1.0f;
+
+		} else if (_vtol_schedule.flight_mode == FW_MODE) {
+			// transition to mc mode
+			_vtol_schedule.flight_mode = TRANSITION_TO_MC;
+			_flag_enable_mc_motors = true;
+			_vtol_schedule.transition_start = hrt_absolute_time();
+
+		} else if (_vtol_schedule.flight_mode == TRANSITION_TO_FW) {
+			// failsafe back to mc mode
+			_vtol_schedule.flight_mode = MC_MODE;
+
+		} else if (_vtol_schedule.flight_mode == TRANSITION_TO_MC) {
+			// keep transitioning to mc mode
+			_vtol_schedule.flight_mode = MC_MODE;
+		}
+
+		// the pusher motor should never be powered when in or transitioning to mc mode
+		_pusher_throttle = 0.0f;
+
+	} else {
+		// the transition to fw mode switch is on
+		if (_vtol_schedule.flight_mode == MC_MODE) {
+			// start transition to fw mode
+			_vtol_schedule.flight_mode = TRANSITION_TO_FW;
+			_pusher_throttle = 0.0f;	// start from zero so we can ramp up from a known min. value
+			_vtol_schedule.transition_start = hrt_absolute_time();
+
+		} else if (_vtol_schedule.flight_mode == FW_MODE) {
+			// in fw mode
+			_vtol_schedule.flight_mode = FW_MODE;
+			// use commanded fw throttle
+			_pusher_throttle = _actuators_fw_in->control[3];
+
+		} else if (_vtol_schedule.flight_mode == TRANSITION_TO_FW) {
+			// continue the transition to fw mode while monitoring airspeed for a final switch to fw mode
+			if (_airspeed->true_airspeed_m_s >= _params_pusher.airspeed_trans) {
+				_vtol_schedule.flight_mode = FW_MODE;
+				// we can turn off the multirotor motors now
+				_flag_enable_mc_motors = false;
+				// don't set pusher throttle here as it's being ramped up elsewhere
+			}
+
+		} else if (_vtol_schedule.flight_mode == TRANSITION_TO_MC) {
+			// transitioning to mc mode & transition switch on - failsafe into fw mode
+			_vtol_schedule.flight_mode = FW_MODE;
+			_pusher_throttle = _actuators_fw_in->control[3];
+		}
+	}
+
+	update_transition_state();
+
+	// map pusher specific control phases to simple control modes
+	switch(_vtol_schedule.flight_mode) {
+		case MC_MODE:
+			_vtol_mode = ROTARY_WING;
+			break;
+		case FW_MODE:
+			_vtol_mode = FIXED_WING;
+			break;
+		case TRANSITION_TO_FW:
+		case TRANSITION_TO_MC:
+			_vtol_mode = TRANSITION;
+			break;
+	}
+}
+
+void Pusher::update_transition_state()
+{
+	if (_vtol_schedule.flight_mode == TRANSITION_TO_FW) {
+		// Ramp up throttle to the current throttle value to something above cruise throttle
+		// NOTE: Probably need to be a bit more clever than this. Should we keep ramping up to full throttle during which time transition airspeed should be reached?
+		if (_pusher_throttle <= _params_pusher.pusher_transition) {
+			_pusher_throttle = _params_pusher.pusher_mc + fabsf(_params_pusher.pusher_transition - _params_pusher.pusher_mc) * 
+						(float)hrt_elapsed_time(&_vtol_schedule.transition_start) / (_params_pusher.front_trans_dur * 1000000.0f);
```

Comment:
But this section as it currently stands has several potential divide by zeros begging to happen, so it still needs correcting.